### ISSUE TITLE
Training Reproducibility

### ISSUE THREAD

**Issue by reginehartwig** [monniert/unicorn#8]
I am currently analyzing the training process of your model.
I recognized that results are only partially reproducible as there seems to be some randomness in the training.

Do you know which parts of the code are influencing the reproducibility? Could it be Pytorch3D, also related to this issue https://github.com/facebookresearch/pytorch3d/issues/659?
It would be great if you could tell me more about it, the tests you might have run, and whether you plan to work on this.

**Comment by monniert** [comment]
Hi @reginehartwig! Which experiments are you having issue to reproduce? As stated in the readme, for complex real images like birds and horses, we observed that the model can still converge to a bad local minima where the prototypical shape is wrong, and you should try another random seed and check the results after the first stage. It is difficult to reproduce exactly this kind of experiments even by setting the random seed, it can depend on the issue you pointed out or it could also depend on the hardware and the versions of librairies you installed.

**Comment by reginehartwig** [comment]
Hi @monniert! Thanks for the fast reply! 
I ran experiments with p3dcar, cub and shapenetnmr. The problem is that I still get different results for multiple runs, even by taking the same random seed and running with the same setting. The plotted loss curves also reflect that already at the first few epochs.
![loss_uniform](https://github.com/monniert/unicorn/assets/28822007/990f333d-5838-4ced-bf2b-85a11f84fb48)

Later on the results can become very different. This means, I cannot run a code twice (with the same seed) and expect the same outcome.
![loss_uniform_500_epochs](https://github.com/monniert/unicorn/assets/28822007/2dc99b7b-ae5a-4851-8114-91ad1b203135)
![chamfer-l1-icp](https://github.com/monniert/unicorn/assets/28822007/ee9a1ec2-adc6-4491-a405-bd23560ac067)

**Comment by monniert** [comment]
From what I remember in my case, I think beginning of trainings were mostly identical by fixing the seed, not really sure about performances in the long run though. Are you always running the experiments on the same machine? The source of randomness can come from different tiny things, you should investigate the common source of randomness listed at this link (https://pytorch.org/docs/stable/notes/randomness.html), in particular you should set `torch.backends.cudnn.benchmark = False` L27 in `src.trainer.py`

It could also be related to the issue you mentioned. I do not plan to work on this, but would be interested to hear about the root cause if you manage to make it completely deterministic
